docs(api): add documentation for multiples of a module (#7525)
OpenTrons/opentrons-api: api/docs/v2/new_modules.rst

  tc_mod.deactivate_block()

.. versionadded:: 2.0

***************************************
Using Multiple Modules of the Same Type
***************************************

The software now supports multiple modules of the same type in a protocol. To use this feature, you must be connected to
the robot via the Opentrons Run App and be running software version 4.3 or higher. Support for multiple modules is only

> available for either temperature modules or magnetic modules at this time.

Reviewer: > The software now supports multiple modules of the same type in a protocol. To use this feature, you must be connected to the robot via the Opentrons Run App and be running software version 4.3 or higher.

* I think you can remove the phrase "the software now supports...". Generally, I think it's a better fit for release notes and version change notes. Here, we want to describe the state of the current system to an audience for which the phrase "the software now supports" isn't necessarily meaningful.
* "Opentrons Run App" should be "Opentrons App," per style guidance to avoid the term "run app" in customer-facing documentation.
* What's the significance of connecting to the robot via the Opentrons App? What's the alternative? Jupyter?

-----

> Support for multiple modules is only available for either temperature modules or magnetic modules at this time.

Would it be clearer to rephrase this as:

> Currently, you can only use multiple Magnetic Modules or multiple Temperature Modules. You won't be able to load multiple Thermocycler Modules.

-----

* Wherever module names are mentioned, they generally should be capitalized ("Temperature Module," "Magnetic Module"), since they're products. (Per style guidance.)

> To use multiple modules, simply load two modules of the same type into your protocol. The robot will then map out

Reviewer 2: ```suggestion
To use multiple modules, load two modules of the same type into your protocol. The robot will then map out
```

> each module in order of appearance to the physical robot ports in order of appearance. That means if you load 

Reviewer: Will the Opentrons App visually guide users through plugging modules in specific deck slots into specific USB ports? How much of this do we need to explain here?

If we do need to explain it here, before going into the details, can we back up a little and introduce users to what the problem is and why they need to care about them?

Reviewer: The tension in the other direction is that these concepts might be shared with Protocol Designer's implementation of MoaM, so we shouldn't necessarily explain them at all here in the API docs. The alternatives would be in-app copy or Help Center articles.

I haven't been keeping up with how this feature will look in Protocol Designer or the Opentrons App, so I don't have an informed opinion on how to deal with that potential duplication.

Author: I am planning on holding a documentation meeting with support. Maybe I can hold off on merging this until I meet with them. I don't know that we should be explaining what the problem is here -- to me API documentation serves as 'hey we have this cool feature, this is how it works' not necessarily why we have this feature.

I do agree we want to make sure we're not duplicating info in multiple places, so the docs change is something we can go over in the meeting.

two temperature modules, the temperature module attached to the left-most port will be related to the first temperature
module in your protocol while the second temperature module loaded would be related to the temperature module connected to the

> right-most port.

Reviewer: ```suggestion
next port to the right.
```

I think?

The following diagram shows the mapping of two temperature modules on the robot.

.. image:: ../img/modules/multiples_of_a_module.png

In a protocol, the diagram would map to your modules as found below.

.. code-block:: python

>     :substitutions:

Reviewer: Is this `:substitutions:` here accidentally or is it a weird Sphinx/reStructuredText thing?

Author: This allows the api level to change dynamically every time the value is incremented in the code. See: https://github.com/Opentrons/opentrons/pull/6292/files

    from opentrons import protocol_api

    metadata = {'apiLevel': '|apiLevel|'}

    def run(protocol: protocol_api.ProtocolContext):
         # Load Temperature Module A in deck slot 1 on port 1 of the robot.

>          temperature_module_a = protocol.load_module('temperature module gen2', 1)

Reviewer: Should we suggest providing a `label` to `load_module()`? Would that show up in the Opentrons App in a useful way?

Author: We sorta removed labels as a concept from API v2. I think we would need to think more about what would make a label actually useful to a user. Right now, they would use the slot numbers as guidance in the app, which I think is sufficient.

         # Load Temperature Module B in deck slot 3 on port 1 of the hub.
         temperature_module_b = protocol.load_module('temperature module gen2', 3)

Referencing the diagram, you should make sure that Temperature Module A in slot 1 is plugged into
port 1 of the robot and Temperature Module B in slot 3 is plugged into port 1 of the hub.

If for whatever reason you want to plug Temperature Module B into a port to the left of Temperature Module A,

> you should simply switch the order that you call the modules.

Reviewer 2: ```suggestion
you must switch the order that you call the modules.
```

> you should simply switch the order that you call the modules.

Reviewer: Nit:

```suggestion
you should switch the order in which you load the modules.
```

Or, to be a little more explicit:

```suggestion
you should load Temperature Module B before Temperature Module A.
```

         # Load Temperature Module A in deck slot 1 on port 1 of the hub.
         temperature_module_a = protocol.load_module('temperature module gen2', 1)

> The Opentrons App will display USB Port information while using modules in your protocol, however, you can use the
> guidelines from above to also use this feature in jupyter notebook.

Reviewer: ```suggestion
The Opentrons App will display USB Port information while using modules in your protocol. However, you can use the
guidelines above to get a predictable module order in Jupyter Notebook.
```

Micro comments:

* `,` -> `.` because I think that's a comma splice.
* `jupyter notebook` -> `Jupyter Notebook`
* `use this feature` -> `get a predictable module order` if you think that makes this clearer?

Reviewer: Macro comments, which I think are more important, but I understand if you want to punt them to another PR:

It seems like we spend a lot of space explaining the details of how this feature works. But then, down here, at the very end, we imply that those details aren't critical to know unless you're using Jupyter Notebook.

I'm not up to speed on this feature's UX, so you'll know better than I. But, I suspect that the details *aren't* critical. They're important to document, but a user doesn't *really* have to understand all of them before they can successfully use multiples of a module, right?

With that in mind, would an outline like this work?

1. "Since API version __, a single protocol can control multiple hardware modules of the same type, like two Temperature Modules."
2. [Brief code sample showing what we mean.]
    * *Putting the code sample high up keeps things concrete and easy to follow.*
3. "The OT-2 might need your help to identify which hardware module in your code corresponds to which hardware module in the physical world."
    * *In another thread, we discussed whether it's appropriate for the API docs to explain the background of this problem. IMO a brief, high-level sentence like this is the best of both worlds, but what do you think?*
4. "If you're uploading your protocol through the Opentrons App, it will guide you through confirming the mapping before you run it."
    * *This is placed high up to act as an early escape hatch for people to decide they don't care, before they get in too deep.*
5. "However, you can also control the mapping by changing the order of your `load_module()` calls. This can be useful if you're using the Python Protocol API your OT-2 through Jupyter Notebook. [Link to Jupyter Notebook docs.]
    * *Brief intro to what we're about to explain before we even start explaining it. Again, this comes up-front to help readers decide whether or not they care to read further.*
6. [Details about how the mapping works.]

This is just a sketch; the exact phrasing can change, of course.

Reviewer: @Laura-Danielle and I talked about my macro comments in Slack.

Under the current designs (https://github.com/Opentrons/opentrons/pull/7543), the app doesn't exactly "guide users" to plug the modules into the right ports. It's more like the app *informs* them how the robot happens to have mapped the modules.

Laura explained it like (emphasis mine):

> user goes to labware calibration screen and they see which modules are plugged in or not. If a module is plugged in, the app displays which module that maps to in the protocol (i.e. temp mod in slot one). **If the modules are plugged in incorrectly, they have no path to resolve this other than understanding that the modules need to be plugged in from left to right based on appearance order in the protocol.**

I do still think that the explication in this PR is a little heavy, but I also see why we feel like that heaviness is warranted. I'm going to keep looking for ways to make it more minimal while still conveying what's necessary for users to correctly run their multiples-of-a-module protocols. For this PR, though, we're ignoring my macro comments.

-----

My micro comments still apply. I think this PR is good to merge with those points addressed. Approving it now for the sake of streamlining.
